packages/tre/src/plugins/kv/remote.ts

```diff
@@ -0,0 +1,273 @@
+  actualExpiration?: number;
+  // The actual user-specified metadata of this key, if any. We want to return
+  // this to users instead of this object.
+  actualMetadata?: unknown;
+}
+
+const APIEnvelopeSchema = z.object({
+  success: z.boolean(),
+  errors: z.array(z.object({ code: z.number(), message: z.string() })),
+  messages: z.array(z.object({ code: z.number(), message: z.string() })),
+});
+
+const KVGetMetadataResponseSchema = z.intersection(
+  APIEnvelopeSchema,
+  z.object({ result: z.unknown() })
+);
+
+const KVListResponseSchema = z.intersection(
+  APIEnvelopeSchema,
+  z.object({
+    result: z.array(
+      z.object({
+        name: z.string(),
+        expiration: z.onumber(),
+        metadata: z.unknown(),
+      })
+    ),
+    result_info: z.optional(
+      z.object({
+        count: z.onumber(),
+        cursor: z.ostring(),
+      })
+    ),
+  })
+);
+
+async function assertSuccessfulResponse(response: Response) {
+  if (response.ok) return;
+
+  // If this wasn't a successful response, throw a KVError
+  const contentType = response.headers.get("Content-Type");
+  if (contentType?.toLowerCase().includes("application/json")) {
+    const envelope = APIEnvelopeSchema.parse(await response.json());
+    throw new KVError(
+      response.status,
+      envelope.errors.map(({ message }) => message).join("\n")
+    );
+  } else {
+    throw new KVError(response.status, await response.text());
+  }
+}
+
+const DEFAULT_CACHE_TTL = 60;
+// Returns seconds since UNIX epoch key should expire, using the specified
+// expiration only if it is sooner than the cache TTL
+function getCacheExpiration(
+  clock: Clock,
+  expiration?: number,
+  cacheTtl = DEFAULT_CACHE_TTL
+): number {
+  // Return minimum expiration
+  const cacheExpiration = millisToSeconds(clock()) + cacheTtl;
+  if (expiration === undefined || isNaN(expiration)) return cacheExpiration;
+  else return Math.min(cacheExpiration, expiration);
+}
+
+function cacheNeedsRevalidation(
+  clock: Clock,
+  cacheTtl: number,
+  value: StoredValueMeta<RemoteCacheMetadata>
+): boolean {
+  // If according to the new cache TTL, the key should've expired already,
+  // revalidate it
+  if (value.metadata?.storedAt !== undefined) {
+    const newExpiration = value.metadata.storedAt + cacheTtl;
+    return newExpiration < millisToSeconds(clock());
+  }
+  return false;
+}
+
+export class KVRemoteStorage extends RemoteStorage {
+  async get(
+    key: string,
+    skipMetadata?: boolean,
+    cacheTtl = DEFAULT_CACHE_TTL
+  ): Promise<StoredValueMeta | undefined> {
+    // If this key is cached, return it
+    const cachedValue = await this.cache.get<RemoteCacheMetadata>(key);
+    // cacheTtl may have changed between the original get call that cached
+    // this value and now, so check the cache is still fresh with the new TTL
+    if (
+      cachedValue?.metadata !== undefined &&
+      !cacheNeedsRevalidation(this.clock, cacheTtl, cachedValue)
```

Comment:
I'm not certain, but from reading the docs I think we'd also need to _store_ this new `cacheTtl`